src/com/amazon/ionelement/api/Ion.kt

```diff
@@ -25,134 +27,141 @@ import com.amazon.ionelement.impl.DecimalIonElement
-fun ionSexpOf(iterable: Iterable<IonElement>, metas: MetaContainer = emptyMetaContainer()): IonElementContainer =
+/** Creates a [SexpElement] that represents an Ion `sexp`.*/
+fun ionSexpOf(iterable: Iterable<Element>, metas: MetaContainer = emptyMetaContainer()): SexpElement =
     when {
         iterable.none() -> EMPTY_SEXP
-        else -> SexpIonElementArray(iterable.toList(), metas = metas)
+        else -> SexpIonElementArray(iterable.map { it as IonElement }, metas = metas)
     }
 
-/** Creates an [IonElement] that represents an Ion `sexp`.*/
-fun ionSexpOf(vararg elements: IonElement, metas: MetaContainer = emptyMetaContainer()): IonElementContainer =
+/** Creates a [SexpElement] that represents an Ion `sexp`.*/
+fun ionSexpOf(vararg elements: Element, metas: MetaContainer = emptyMetaContainer()): SexpElement =
     ionSexpOf(elements.asIterable(), metas)
 
-/** Returns an [IonElement] representing an empty Ion `list` .*/
-fun emptyIonList(): IonElementContainer = EMPTY_LIST
+/** Returns a [ListElement] representing an empty Ion `list` .*/
+fun emptyIonList(): ListElement = EMPTY_LIST
 
-/** Returns an [IonElement] representing an empty Ion `sexp` .*/
-fun emptyIonSexp(): IonElementContainer = EMPTY_SEXP
+/** Returns a [SexpElement] representing an empty Ion `sexp` .*/
+fun emptyIonSexp(): SexpElement = EMPTY_SEXP
 
-/** Returns an [IonElement] representing an empty Ion `struct` .*/
-fun emptyIonStruct(): IonElement = EMPTY_STRUCT
+/** Returns a [StructElement] representing an empty Ion `struct` .*/
+fun emptyIonStruct(): StructElement = EMPTY_STRUCT
 
 /** Creates an [IonStructField] .*/
-fun field(key: String, value: IonElement): IonStructField =
-    IonStructFieldImpl(key, value)
+fun field(key: String, value: Element): IonStructField =
+    IonStructFieldImpl(key, value as IonElement)
 
-/** Creates an [IonElement] that represents an Ion `struct` with the specified fields. */
-fun ionStructOf(fields: Iterable<IonStructField>): IonElement =
+/** Creates a [StructElement] that represents an Ion `struct` with the specified fields. */
+fun ionStructOf(fields: Iterable<IonStructField>): StructElement =
     when {
         fields.none() -> EMPTY_STRUCT
         else -> StructIonElementImpl(fields.toList())
     }
 
-/** Creates an [IonElement] that represents an Ion `struct` with the specified fields. */
-fun ionStructOf(vararg fields: IonStructField): IonElement =
+/** Creates a [StructElement] that represents an Ion `struct` with the specified fields. */
+fun ionStructOf(vararg fields: IonStructField): StructElement =
     ionStructOf(fields.asIterable())
 
 /** Creates an [IonElement] that represents an Ion `struct` with the specified fields. */
-fun ionStructOf(vararg fields: Pair<String, IonElement>): IonElement =
-    ionStructOf(fields.map { field(it.first, it.second) })
+fun ionStructOf(vararg fields: Pair<String, Element>): StructElement =
+    ionStructOf(fields.map { field(it.first, it.second as IonElement) })
```

Comment:
We talked about this--but I want to emphasize that I'd rather model this as an explicit API in `Element` versus relying on implementations to implement both (which is not statically enforceable).

Reply:
Replacing casts with `.asIonElement()`.